Doctests and docstrings added

dnabanita7 opened the pull request:
#### References to other Issues or PRs
<!-- Fixes #15370 -->


#### Brief description of what is fixed or changed
I have added parameters,returns and examples sections to series.series in series.py

#### Release Notes


<!-- BEGIN RELEASE NOTES -->
* series
  * series.py
<!-- END RELEASE NOTES -->

dnabanita7 committed:
Changed subs() in matrices for iterable objects

The zip object is iterable.subs method in common.py didn't have a case for iterable objects.
Still I have to change the subs method in core.basic. The isinstance doesn't track iterable objects which don't have _getitem_ module,zip not having one creates an error. So,Shall I change isinstance() check to iter() in subs module in core.basic.

Almost Fixes #10589

I have added tests for the same.

dnabanita7 committed:
Update test_commonmatrix.py

dnabanita7 committed:
Update common.py

dnabanita7 committed:
Added docstrings and doctests

Added paramets,examples and returns for series.py

Sc0rpi0n101 commented on sympy/matrices/common.py:
```diff
@@ -1888,6 +1888,7 @@ def subs(self, *args, **kwargs):  # should mirror core.basic.subs
         """
         return self.applyfunc(lambda x: x.subs(*args, **kwargs))
 
+
```
Only a blank line addition in this file.
Can you remove it?

Naba7 replied:
yes .

Sc0rpi0n101 commented on sympy/series/series.py:
```diff
@@ -4,9 +4,55 @@
 
 def series(expr, x=None, x0=0, n=6, dir="+"):
-    """Series expansion of expr around point `x = x0`.
+    """Returns
```
Might be a good idea to add a description what the function does before this heading instead of starting with what it returns.

Naba7 replied:
Yes,I thought so. Then it returns the same thing as it is described as,so changed it. Although,now I have changed it.

Sc0rpi0n101 commented on sympy/series/series.py:
```diff
@@ -4,9 +4,55 @@
+       x    : Symbol
+              It is the variable of the expression to be calculated.
+
+       x0   : Value
```
```suggestion
       x0   : Integer
```
Integer might be better than `Value`

Naba7 replied:
Well, it can take up decimal values too.

Sc0rpi0n101 replied:
Then, maybe try something like `Number`.
`Value` doesn't  exactly explain a type imo.

Naba7 replied:
Yes, I thought about it too. Alright, I am changing this to ``Number``

Sc0rpi0n101 commented on sympy/series/series.py:
```diff
@@ -4,9 +4,55 @@
+              The value around which ``x`` is calculated. Can be ``oo`` or
+              ``-oo``.
+
+       n    : Value
```
```suggestion
       n    : Integer
```

Naba7 replied:
for n ,i think changing to integer will be no problem.

Naba7 replied:
Infinity object can't be a number.

Sc0rpi0n101 replied:
Yeah, but `oo` doesn't work. It gives a TypeError,
And anything between `-oo` and `oo` can be a number.

A third opinion fro this might be a good idea.

Sc0rpi0n101 commented on sympy/series/series.py:
```diff
@@ -4,9 +4,55 @@
+              It is the variable of the expression to be calculated.
+
+       x0   : Value
+              The value around which ``x`` is calculated. Can be ``oo`` or
```
Can it only be `-oo` or `oo`?
Your examples say otherwise.
Did you mean it can be between `-oo to oo` ?

Naba7 replied:
Oh! yes My bad. I am changing it.

Sc0rpi0n101 commented on sympy/series/series.py:
```diff
@@ -4,9 +4,55 @@
+       >>> from sympy import Symbol, series, tan, oo
+       >>> from sympy.abc import x
+       >>> f = tan(x)
+       >>> f.series(x, 2, 6, "+")
```
It looks like it calls `series(self, x=None, x0=0, n=6, dir="+", logx=None)` 
instead of `series(expr, x=None, x0=0, n=6, dir="+")`
which is `expr.series()`

Similarly, the next examples.

Naba7 replied:
``f `` is an expression and ``f.series()`` is similar to ``expr.series()``

Sc0rpi0n101 replied:
But isn't this function supposed to something like `series(f, x, 2, 6, "+")` ?

Naba7 replied:
I have changed it.

Sc0rpi0n101 commented:
The function acts as a wrapper, it shouldn't require that much explanation.

Also, from what I can understand, the function just simplifies the `expr` and calls `Expr.series()` on it. 

Can you tell me why this function is really required?

Sc0rpi0n101 commented on sympy/series/series.py:
```diff
@@ -4,9 +4,55 @@
+       + O((x - 2)**3, (x, 2))
+
+       >>> f.series(x, 2, oo, "+")
+       Traceback (most recent call last):
```
I don't think exceptions work this way for doctests.

Naba7 replied:
Well, I followed this https://github.com/sympy/sympy/blob/master/sympy/series/limits.py

Sc0rpi0n101 replied:
ohh. it does. Good to know.

Well, the build report report shows  `TypeError`
You can try changing the Error acc. to thr report, that might work.

Naba7 replied:
yes.

sylee957 replied:
This still fails the doctest now. Maybe change like this
https://travis-ci.org/sympy/sympy/jobs/530437132#L754

Sc0rpi0n101 replied:
I don't think it has been changed yet .
Can you try changing it @Naba7 so we can see there are no other problems.

Naba7 replied:
I have changed it now.

Sc0rpi0n101 commented:
Wait, why does this PR have 3 unrelated commits from Mar 28?

You can drop these commits using `rebase`

dnabanita7 committed:
Added docstrings and doctests

in series.py

Sc0rpi0n101 commented:
Also, The release notes are written with an end-user perspective.
I don't think those release notes are really helpful.

Naba7 commented:
> The function acts as a wrapper, it shouldn't require that much explanation.
Well, I think every function and wrapper class should have a detailed explanation for the new-comers to understand without much googling stuff.
> 
> Also, from what I can understand, the function just simplifies the `expr` and calls `Expr.series()` on it.
Well,actually yes. Do you think I should abandon the changes made?
> 
> Can you tell me, is this function really required? If yes, why?
It is the way to call ``.series()`` function, it should have a clear explanation of what it is doing.

dnabanita7 committed:
Added docstrings and doctests

in series.py

Sc0rpi0n101 commented:
> It is the way to call `.series()` function, it should have a clear explanation of what it is doing.

My question is, if this function was removed, wouldn't `f.series(x, 2, 6, "+")` still work?

Sc0rpi0n101 commented:
> Well,actually yes. Do you think I should abandon the changes made?

I'd take someone else's opinion first; as i'm not exactly sure what the purpose of this function is.

Naba7 commented:
> > It is the way to call `.series()` function, it should have a clear explanation of what it is doing.
> 
> My question is, if this function was removed, wouldn't `f.series(x, 2, 6, "+")` still work?

let me try it then. I am not sure about it.

oscargus commented on sympy/series/series.py:
```diff
@@ -6,7 +6,56 @@
+       Examples
+       ========
+
+       >>> from sympy import Symbol, series, tan, oo
```
If you are calling it as `f.series` there is no need to9 import series. On the other hand, if the documentation entry is for the wrapper it makes sense to call it using `series(f, ...)`.

Naba7 replied:
yes,ok.

oscargus commented on sympy/series/series.py:
```diff
@@ -6,7 +6,56 @@
+       =======
+
+       Series expansion of the expression about x0
+       See the docstring of Expr.series() for complete details of this wrapper.
```
It would make sense to add a "See Also" section for this.

Naba7 replied:
Alright  Adding it now.

Naba7 replied:
alright

sylee957 commented on sympy/series/series.py:
```diff
@@ -6,7 +6,56 @@
     """Series expansion of expr around point `x = x0`.
 
-    See the docstring of Expr.series() for complete details of this wrapper.
+       Parameters
```
It is better to dedent all the following lines by 3-spaces.
We generally expect every things to be indented 4*n characters

Naba7 replied:
I am sorry. I couldn't understand you.

Sc0rpi0n101 replied:
He meant to say, you have to move all your changes back 3 spaces.
The text should be in line with the first `"`

```python
"""This is the heading

Rest of the test should be indented like this
   Not like this 
```

Naba7 replied:
Alright. fine. Thanks.

Naba7 replied:
What about 4*n characters?

Sc0rpi0n101 replied:
The python code conventions state that every indent should be exactly 4 characters
The Parameters in your code starts at `4+4+3 = 11` characters which violates the conventions.
So, it is better to start the text at same indent as `"` which is `4+4` characters here

Naba7 replied:
ok. thanks.

sylee957 reviewed (commented):
<img src=https://user-images.githubusercontent.com/34944973/57545222-a3161c00-7394-11e9-86ec-7d562dfa1b6a.png width=384px>

I think the doc is not built correctly right now.
Maybe a blank line should be added before the ``Examples`` section begins

dnabanita7 committed:
Added doctests and docstrings

in series.py

dnabanita7 committed:
Added docstrings and doctests

changed indentation

Naba7 commented:
Is it ready to go?  @oscargus @Sc0rpi0n101 @sylee957

sylee957 commented:
After seeing how the doc of the wrapped `Expr.series` looks like, I think that this doc may better go to the core module
https://docs.sympy.org/dev/modules/core.html#sympy.core.expr.Expr.series

Naba7 commented on sympy/series/series.py:
```diff
@@ -6,6 +6,72 @@
+        s1 = self._eval_nseries(x, n=n, logx=logx)
+      File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/sympy-1.5.dev0-py3.7.egg/sympy/functions/elementary/trigonometric.py", line 1094, in _eval_nseries
+        return Function._eval_nseries(self, x, n=n, logx=logx)
+      File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/sympy-1.5.dev0-py3.7.egg/sympy/core/function.py", line 707, in _eval_nseries
```
No,from sympy import Symbol, series,tan, oo it doesn't. I was confused if I would copy the contents in my terminal or the ci. Shall I change it to the output of my terminal?

sylee957 replied:
It is generally not a good idea to copy-paste the full traceback log, as it may contain specific information about your local file system.
If you copy paste your own log from your terminal, it may not work with the CI or other system.

Maybe see how the test like this implemented in `limits`
https://github.com/sympy/sympy/blob/be5a4725133f29b22e5a7e8a3d04ecbc969f1fe7/sympy/series/limits.py#L50-L53

Naba7 commented on sympy/series/series.py:
```diff
@@ -6,6 +6,72 @@
+           The expression whose series is to be expanded.
+
+    x    : Symbol
+           It is the variable of the expression to be calculated.
```
Alright then.

Naba7 replied:
I placed the colon symbol in an arranged line.

Naba7 replied:
I have changed this.

Naba7 commented on sympy/series/series.py:
```diff
@@ -6,6 +6,72 @@
+           The number of terms upto which the series is to be expanded.
+
+    dir  : String
+           Optional (default: "+")
```
ok,thanks

Naba7 commented on sympy/series/series.py:
```diff
@@ -6,6 +6,72 @@
+    Returns
+    =======
+
+    Series expansion of the expression about x0
```
Ok,changing this.

Naba7 commented:
Shall I make changes in the core module or keep the changes in here?

sylee957 commented:
Yes, please make changes in the core module, as incrementally as possible.

Naba7 commented:
Ok

dnabanita7 committed:
Changed the documentation

Added parameters, returns and examples section in sympy/core/expr.py under series.

czgdp1807 commented on sympy/core/expr.py:
```diff
@@ -2673,6 +2703,41 @@ def series(self, x=None, x0=0, n=6, dir="+", logx=None):
+        tan(2) + (2 - x)*(-tan(2)**2 - 1) + (2 - x)**2*(tan(2)**3 + tan(2))
+        + O((x - 2)**3, (x, 2))
+
+        >>> f.series(x, 2, oo, "+")
```
Can this failed example be included in tests rather than in the docstrings?

Naba7 replied:
Shall I add it in tests and cancel from docstrings?

czgdp1807 replied:
From my point of view, the answer is Yes.

Sc0rpi0n101 replied:
> Shall I add it in tests and cancel from docstrings?

probably, no use of keeping two copies.
it is getting tested both ways.

czgdp1807 replied:
Tests for incorrect inputs should be avoided in docstrings.

Naba7 replied:
alright. How to check typeerror in tests? Shall I use decorator @XFAIL ?

Sc0rpi0n101 replied:
probably
```python
raises(TypeError, lambda: function)
```

sylee957 replied:
In my opinion, it you should make it documented in words if the the test is moved, 
like in `dir` argument, or you can make a new section `Raises` and add possible exceptions that can be thrown from the method.

Sc0rpi0n101 replied:
I like the idea of the Raises section
It can include every possible exception and why were the exceptions raised.
That might be helpful to someone who might not be completely sure why the function is raising an exception.
Although the error messages also include a description of the error but a well documented explanation of why it was raised and what can be done to avoid that can be helpful.

Naba7 replied:
Shall I add this ``raises `` section under def series in sympy/core/expr.py?

Naba7 replied:
IMO, Adding a single line like ``oo`` or ``-oo`` can't be an integer as under ``x0`` and ``n`` will do?

Naba7 replied:
@sylee957 @Sc0rpi0n101 is it good?

czgdp1807 commented on sympy/series/series.py:
```diff
@@ -6,6 +6,72 @@
+    tan(2) + (2 - x)*(-tan(2)**2 - 1) + (2 - x)**2*(tan(2)**3 + tan(2))
+    + O((x - 2)**3, (x, 2))
+
+    >>> series(f, x, 2, oo, "+")
```
> Can this failed example be included in tests rather than in the docstrings?

+1

czgdp1807 commented on sympy/series/series.py:
```diff
@@ -6,6 +6,72 @@
+    =======
+
+    Expr
+        Series expansion of the expression about x0
```
Can this be written as, `Expr : Series expansion of the expression about x0`?

Naba7 replied:
Yes, I am changing .

czgdp1807 commented on sympy/core/expr.py:
```diff
@@ -2673,6 +2703,41 @@ def series(self, x=None, x0=0, n=6, dir="+", logx=None):
+        =======
+
+        Expr
+            Series expansion of the expression about x0
```
> Can this be written as, `Expr : Series expansion of the expression about x0?`

+1

sylee957 commented on sympy/core/expr.py:
```diff
@@ -2645,7 +2645,37 @@ def series(self, x=None, x0=0, n=6, dir="+", logx=None):
+        ==========
+
+        expr : Expression
+               The expression whose series is to be expanded.
```
I'd expect it to be indented 4*n characters, and some below
```suggestion
            The expression whose series is to be expanded.
```

Naba7 replied:
Its okay. I checked the numpydoc format now.

dnabanita7 committed:
Added tests and a raises section

shifted the doctest to tests

oscarbenjamin commented on sympy/series/series.py:
```diff
@@ -6,6 +6,72 @@
+        return Function._eval_nseries(self, x, n=n, logx=logx)
+      File "/home/chad7/anaconda3/lib/python3.7/site-packages/sympy/core/function.py", line 680, in _eval_nseries
+        for i in range(n - 1):
+    TypeError: 'Infinity' object cannot be interpreted as an integer
```
Can you trim this traceback to just the first and last lines and a `...` in between?

Naba7 replied:
I had done it earlier but then changed to this. Shall I change it?

oscarbenjamin replied:
Yes. At the moment the traceback has unnecessary information including the location of files on your computer.

oscarbenjamin commented on sympy/core/tests/test_expr.py:
```diff
@@ -1832,3 +1832,8 @@ def test_normal():
+
+def test_expr():
+    x = symbols('x')
+    raises(TypeError, lambda: tan(x).series(x, 2, oo, "+"))
```
There is now a merge conflict for these changes (presumably because someone else added to the end of this file).

Can you resolve that by moving this to test_series.py? Maybe there is already a test function there that you can add to. Otherwise a better name is needed than `test_expr` which is not very specific.

oscarbenjamin replied:
Looks like @smichr resolved the merge conflicts

smichr committed:
Merge branch 'master' into doc

oscarbenjamin commented:
I think that shortening the traceback in the doctest is the only thing that needs doing. Otherwise this can be merged

Naba7 commented:
I have messed up with the branches. Can you please merge it ? I will open another issue regarding this and shorten the traceback?

oscarbenjamin committed:
Shorten traceback in doctest

oscarbenjamin commented:
I fixed it by editing in the Github web UI.

Thanks for this. I'll merge now.
